src/runtime_src/core/tools/xbmgmt2/SubCmdProgram.cpp

```diff
@@ -25,17 +25,404 @@ namespace XBU = XBUtilities;
+
+  pri = std::make_unique<firmwareImage>(primary.c_str(), MCS_FIRMWARE_PRIMARY);
+  if (pri->fail())
+    throw xrt_core::error(boost::str(boost::format("Failed to read %s") % primary.c_str()));
+  if (!secondary.empty()) {
+    sec = std::make_unique<firmwareImage>(secondary.c_str(),
+      MCS_FIRMWARE_SECONDARY);
+    if (sec->fail())
+      sec = nullptr;
+  }
+
+  flasher.upgradeFirmware(flashType, pri.get(), sec.get());
+  std::cout << boost::format("%-8s : %s \n") % "INFO" % "Shell is updated succesfully.";
+}
+
+/*
+ * Update SC firmware on the board
+ */
+static void 
+update_SC(uint16_t index, const std::string& file)
+{
+  Flasher flasher(index);
+  if(!flasher.isValid())
+    throw xrt_core::error(boost::str(boost::format("%d is an invalid index") % index));
+
+  std::shared_ptr<firmwareImage> bmc =
+    std::make_shared<firmwareImage>(file.c_str(), BMC_FIRMWARE);
+  if (bmc->fail())
+    throw xrt_core::error(boost::str(boost::format("Failed to read %s") % file.c_str()));
+
+  flasher.upgradeBMCFirmware(bmc.get());
+}
+
+/* 
+ * Helper function for header info
+ */
+static std::string 
+file_size(const char* file) 
+{
+  std::ifstream in(file, std::ifstream::ate | std::ifstream::binary);
+  auto total_size = std::to_string(in.tellg()); 
+  int strSize = static_cast<int>(total_size.size());
+
+  if (strSize > 3) {
+   for (int i = (strSize - 3);  i > 0;  i -= 3)
+     total_size.insert(static_cast<size_t>(i), 1, ',');
+  }
+  
+  return total_size + std::string(" bytes");
+}
+
+/* 
+ * Helper function for header info
+ */
+static std::pair<std::string, std::string> 
+deployment_path_and_filename(std::string file)
+{
+  using tokenizer = boost::tokenizer< boost::char_separator<char> >;
+  boost::char_separator<char> sep("\\/");
+  tokenizer tokens(file, sep);
+  std::string dsafile = "";
+  for (auto tok_iter = tokens.begin(); tok_iter != tokens.end(); ++tok_iter) {
+  	if ((std::string(*tok_iter).find(XSABIN_FILE_SUFFIX) != std::string::npos) 
+          || (std::string(*tok_iter).find(DSABIN_FILE_SUFFIX) != std::string::npos))
+          dsafile = *tok_iter;
+  }
+  auto pos = file.rfind('/') != std::string::npos ? file.rfind('/') : file.rfind('\\');
+  std::string path = file.erase(pos);
+
+  return std::make_pair(dsafile, path);
+}
+
+/* 
+ * Helper function for header info
+ */
+static std::string 
+get_file_timestamp(const char* file) 
+{
+  std::time_t ftime = boost::filesystem::last_write_time(boost::filesystem::path(file));
+  return std::string(std::asctime(std::localtime(&ftime)));
+}
+
+/* 
+ * Find the correct shell to be flashed on the board
+ * Helper method for auto_flash
+ */
+static DSAInfo 
+selectShell(uint16_t idx, const std::string& dsa, const std::string& id)
+{
+  uint16_t candidateDSAIndex = std::numeric_limits<uint16_t>::max();
+  Flasher flasher(idx);
+  if(!flasher.isValid())
+    throw xrt_core::error(boost::str(boost::format("%d is an invalid index") % idx));
+
```

Comment:
Yes. that sounds like a good idea. I'll expose the function from flasher.h. 
You're right. Looks like ID (platform ID) should be unsigned long, but It's being used as string. How about I push this to the next PR since a lot of functions are tied to ID